File: csharp/935.knight-dialer.cs
/*
 * @lc app=leetcode id=935 lang=csharp
 *
 * [935] Knight Dialer
 */

// @lc code=start
public class Solution {

    private static int[][] Paths = new int[][] {
        new int[] { 4, 6 },
        new int[] { 6, 8 },
        new int[] { 7, 9 },
        new int[] { 4, 8 },
        new int[] { 0, 3, 9 },
        new int[] {},
        new int[] { 0, 1, 7 },
        new int[] { 2, 6 },
        new int[] { 1, 9 },
        new int[] { 2, 4 }
    };

    public int KnightDialer(int N) {
        int r = 0, ret = 0;
        int[,] dp = new int[2, 10];
        for (int j = 0; j < 10; ++j) {
            dp[1, j] = 1;
        }
        for (int i = 1; i < N; ++i) {
            for (int j = 0; j < 10; ++j) {
                dp[r, j] = 0;
                foreach (int k in Paths[j]) {
                    dp[r, j] = (dp[r, j] + dp[1 - r, k]) % 1000000007;
                }
            }
            r = 1 - r;
        }
        for (int j = 0; j < 10; ++j) {
            ret = (ret + dp[1 - r, j]) % 1000000007;
        }
        return ret;
    }
}
// @lc code=end

File: java/935.knight-dialer.java
/*
 * @lc app=leetcode id=935 lang=java
 *
 * [935] Knight Dialer
 */

import java.util.Collections;
import java.util.Map;
import java.util.Set;

// @lc code=start
class Solution {
    private static final Map<Integer, Set<Integer>> MOVE = Map.of(
        0, Set.of(4, 6),
        1, Set.of(6, 8),
        2, Set.of(7, 9),
        3, Set.of(4, 8),
        4, Set.of(0, 3, 9),
        5, Collections.emptySet(),
        6, Set.of(0, 1, 7),
        7, Set.of(2, 6),
        8, Set.of(1, 3),
        9, Set.of(2, 4)
    );

    public int knightDialer(int n) {
        int[][] dp = new int[2][10];
        int r = 0;
        for (int i = 0; i < 10; ++i) {
            dp[0][i] = 1;
        }
        for (int i = 1; i < n; ++i) {
            r = 1 - r;
            for (int j = 0; j < 10; ++j) {
                dp[r][j] = 0;
                for (int prev : MOVE.get(j)) {
                    dp[r][j] = (dp[r][j] + dp[1 - r][prev]) % 1000000007;
                }
            }
        }
        int ret = 0;
        for (int i = 0; i < 10; ++i) {
            ret = (ret + dp[r][i]) % 1000000007;
        }
        return ret;
    }
}
// @lc code=end

